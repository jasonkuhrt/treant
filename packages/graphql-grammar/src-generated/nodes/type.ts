import type { Node } from 'web-tree-sitter';
import type { ListType } from './list_type.js';
import type { NamedType } from './named_type.js';
import type { NonNullType } from './non_null_type.js';

const TYPE = 'type' as const;

/**
 * Represents a type in the graphql AST.
 */
export interface Type extends Node {
  type: typeof TYPE;
}

/**
 * Type guard for Type
 */
export function isType(node: unknown): node is Type {
  return (node as any)?.type === TYPE;
}