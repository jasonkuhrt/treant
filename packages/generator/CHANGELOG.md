# @treant/generator

## 0.0.1

### Patch Changes

- edd3a65: feat: initial release
- Updated dependencies [edd3a65]
  - @treant/core@0.0.1
  - @treant/grammar@0.0.1
