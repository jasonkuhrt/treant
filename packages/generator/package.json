{
  "name": "@treant/generator",
  "version": "0.0.0",
  "description": "TypeScript SDK generator for tree-sitter grammars",
  "type": "module",
  "author": "Jason Kuhrt",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/jasonkuhrt/treant.git",
    "directory": "packages/generator-sdk"
  },
  "exports": {
    ".": {
      "import": "./build/$.js"
    }
  },
  "files": [
    "build"
  ],
  "scripts": {
    "build": "tsc -p tsconfig.build.json",
    "check:types": "tsc --noEmit",
    "test": "vitest",
    "test:ci": "vitest run",
    "generate:test:fixtures": "tsx tests/helpers/generate-sdk-fixtures"
  },
  "dependencies": {
    "@dprint/formatter": "^0.4.1",
    "@effect/platform": "^0.90.6",
    "@effect/platform-node": "^0.96.0",
    "@treant/core": "workspace:*",
    "@treant/grammar": "workspace:*",
    "effect": "^3.17.9"
  },
  "devDependencies": {
    "web-tree-sitter": "^0.25.8"
  },
  "publishConfig": {
    "access": "public"
  }
}
