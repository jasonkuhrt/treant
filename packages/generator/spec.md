# Introduction

Tree-sitter grammars define AST structure using patterns. Each pattern type generates specific navigator API shapes.

# Grammar Overview

Before diving into patterns, understand these grammar primitives:

- **SYMBOL**: A reference to another rule (like a function call). Doesn't generate API directly - we follow the reference to see what pattern that rule uses.
- **STRING**: A literal string in the grammar (like `"{"` or `"query"`). Usually ignored in navigation.
- **PATTERN/TOKEN**: Regex patterns for lexical analysis. Generate leaf nodes with `.text()` methods.
- **BLANK**: Represents "nothing" - used in CHOICE to make things optional.

- `SEQ` - ordered sequence of elements

## AST Examples

**Optional field (CHOICE with BLANK)**:

```json
{
  "type": "CHOICE",
  "members": [
    { "type": "SYMBOL", "name": "description" },
    { "type": "BLANK" }
  ]
}
```

**Optional with multiple items (CHOICE of REPEAT and BLANK)**:

```json
{
  "type": "CHOICE",
  "members": [
    { "type": "REPEAT", "content": { "type": "SYMBOL", "name": "directive" } },
    { "type": "BLANK" }
  ]
}
```

**Separated list (SEQ with separators)**:

```json
{
  "type": "SEQ",
  "members": [
    { "type": "SYMBOL", "name": "argument" },
    {
      "type": "REPEAT",
      "content": {
        "type": "SEQ",
        "members": [
          { "type": "STRING", "value": "," },
          { "type": "SYMBOL", "name": "argument" }
        ]
      }
    }
  ]
}
```

## Spec Notation Grammar

This specification uses clear unicode notation to describe navigator patterns:

- `«»` - Interface/structure boundaries
- `◆` - Variable parts (member names, child names, parameters)
- `→` - Return type indicator
- `?` - Optional/nullable indicator
- `«◆KEY: $ → ◆R»` - Object/map structure

**Example**: `if◆MEMBER() → Navigator?` means "method named `if` + member name, returns Navigator or null"

## Interface Summary

```
// SEQ Pattern → Child accessor methods
SEQ_Navigator «
  ◆CHILDNAME() → ChildNavigator        // Required child
  ◆CHILDNAME() → ChildNavigator?       // Optional child (CHOICE with BLANK)
»

// CHOICE Pattern → Type narrowing methods
CHOICE_Navigator «
  // Type guards
  is(◆MEMBER) → boolean               // is('field'), is('variable')
  is◆MEMBER() → boolean               // isField(), isVariable()

  // Conditional navigation
  if(◆MEMBER) → Navigator?            // if('field'), if('variable')
  if◆MEMBER() → Navigator?            // ifField(), ifVariable()

  // Pattern matching with transformation
  match(«◆MEMBER: $ → ◆R») → ◆R1 | ◆R2 | ... | null  // $ = navigator for each member
»

// REPEAT/REPEAT1 Pattern → Collection methods
REPEAT_Navigator «
  ◆PLURALNAME() → CollectionNavigator  // document.definitions()
  at(◆N) → Navigator?
  first() / last() → Navigator?
  filter/where/map/reduce/etc.
  [Symbol.iterator]()
»

// All Navigators → Common methods
BASE_Navigator «
  value() → Node?                      // Get underlying node
  text() → string?                     // For leaf nodes
  exists() → boolean
  parent() / ancestor(◆TYPE) → Navigator?
  when(◆PREDICATE) → Navigator?
»

// Terminal Nodes (PATTERN/TOKEN) → Text methods
LEAF_Navigator «
  text() → string                      // name.text() → "userId"
  value() → LeafNode
»
```

Complete Collection API

```
CollectionNavigator «
  // ============= Chain Continuation Methods =============

  // Single item access (returns navigator)
  at(◆INDEX) → Navigator?
  first() → Navigator?
  last() → Navigator?
  find(◆FN) → Navigator?

  // Filtering (returns collection navigator)
  filter(◆FN) → CollectionNavigator
  where(◆predicate) → CollectionNavigator
  whereType(◆type) → CollectionNavigator
  whereName(◆NAME) → Navigator?
  whereText(◆TEXT) → Navigator?

  // Slicing (returns collection navigator)
  slice(◆START, ◆END?) → CollectionNavigator
  take(◆n) → CollectionNavigator
  skip(◆n) → CollectionNavigator

  // ============= Terminal Methods =============

  // Extract values
  value() → ◆T[]                     // Alias for values()
  values() → ◆T[]                    // Get all underlying nodes
  texts() → (string?)[]              // Get all text content

  // Array transformation (terminal)
  map(◆FN) → ◆R[]
  flatMap(◆fn) → ◆R[]
  reduce(◆fn, ◆INITIAL) → ◆R

  // Array queries (terminal)
  some(◆fn) → boolean
  every(◆fn) → boolean
  includes(◆ITEM) → boolean
  indexOf(◆ITEM) → number
  findIndex(◆fn) → number

  // Other terminals
  count() → number
  length → number                     // Alias for count()
  isEmpty() → boolean
  forEach(◆fn) → void
  toArray() → Navigator[]             // Get array of navigators

  // Iterator protocol
  [Symbol.iterator]() → Iterator
»
```

# Interface Mapping

## SEQ

**What it is**: SEQ defines an ordered sequence of child elements. It's like a struct - specific fields in a specific order.

**Grammar Example**:

```json
{
  "field": {
    "type": "SEQ",
    "members": [
      // Optional alias (CHOICE with BLANK = optional)
      {
        "type": "CHOICE",
        "members": [
          { "type": "SYMBOL", "name": "alias" },
          { "type": "BLANK" }
        ]
      },
      // Required name
      { "type": "SYMBOL", "name": "name" },
      // Optional arguments
      {
        "type": "CHOICE",
        "members": [
          { "type": "SYMBOL", "name": "arguments" },
          { "type": "BLANK" }
        ]
      },
      // Optional directives (can be multiple via REPEAT)
      {
        "type": "CHOICE",
        "members": [
          { "type": "SYMBOL", "name": "directives" },
          { "type": "BLANK" }
        ]
      },
      // Optional selection_set
      {
        "type": "CHOICE",
        "members": [
          { "type": "SYMBOL", "name": "selection_set" },
          { "type": "BLANK" }
        ]
      }
    ]
  }
}
```

**What SEQ generates in the Navigator API**:

- One method per child element
- Methods return navigator or null based on optionality
- Method names match the grammar rule names (camelCased)

**Complete API for SEQ pattern**:

```
FieldNavigator «
  // Each member of the SEQ gets a method
  alias() → AliasNavigator?            // Optional (CHOICE with BLANK)
  name() → NameNavigator                // Required (no BLANK option)
  arguments() → ArgumentsNavigator?     // Optional
  directives() → DirectivesNavigator?   // Optional
  selectionSet() → SelectionSetNavigator? // Optional

  // Common navigator methods
  value() → Field
  text() → string?
  parent() → Navigator?
»
```

**Usage Example**:

```typescript
const field = navigate(selection).ifField();

// Access each part of the sequence
const alias = field?.alias(); // Might be null
const name = field?.name(); // Always exists if field exists
const args = field?.arguments(); // Might be null

// Chain through the structure
const fieldName = field?.name().text(); // "userId"
const aliasName = field?.alias()?.name().text(); // "id" or undefined
```

## CHOICE - Union of Possible Grammar Rules

**What it is**: CHOICE means "at this position, we can have one of several different patterns". It's like a union type - the actual node will be exactly one of the options.

**Grammar Example**:

```json
{
  // A 'value' can be many different types
  "value": {
    "type": "CHOICE",
    "members": [
      { "type": "SYMBOL", "name": "variable" },
      { "type": "SYMBOL", "name": "int_value" },
      { "type": "SYMBOL", "name": "float_value" },
      { "type": "SYMBOL", "name": "string_value" },
      { "type": "SYMBOL", "name": "boolean_value" },
      { "type": "SYMBOL", "name": "null_value" },
      { "type": "SYMBOL", "name": "enum_value" },
      { "type": "SYMBOL", "name": "list_value" },
      { "type": "SYMBOL", "name": "object_value" }
    ]
  }
}
```

**What CHOICE generates in the Navigator API**:

1. A base navigator representing the union
2. Type guard methods (`is()`) to check which variant we have
3. Type narrowing methods (`as*()`) to cast to specific types
4. Pattern matching (`match()`) for exhaustive handling

**Complete API for CHOICE pattern**:

```
ValueNavigator «
  // The underlying node is one of the union members
  value() → Value                      // Union type: Variable | IntValue | FloatValue | ...

  // Type guards - check which variant this is
  is(◆MEMBER) → boolean               // is('variable'), is('int_value'), etc.
  is◆MEMBER() → boolean               // isVariable(), isIntValue(), etc.

  // Conditional navigation - returns navigator only if type matches
  if(◆MEMBER) → ◆MEMBERNAVIGATOR?    // if('variable'), if('int_value'), etc.
  if◆MEMBER() → ◆MEMBERNAVIGATOR?    // ifVariable(), ifIntValue(), etc.

  // Pattern matching - keys map to CHOICE member names from grammar
  // Returns union of all handler return types + null (if unmatched member)
  match(«◆member: $ → ◆R, _?: () → ◆R») → ◆R1 | ◆R2 | ... | null
»
```

**Usage Examples**:

```typescript
// Type guard with narrowing
if (value.is('string_value')) {
  console.log(value.text()); // TypeScript knows it's StringValueNavigator
}

// Conditional navigator
const intValue = value.ifIntValue();
if (intValue) {
  console.log(intValue.text());
}

// Pattern matching
const result = value.match({
  variable: v => `$${v.name().text()}`,
  int_value: i => i.text(),
  _: () => 'complex value',
});
```

## REPEAT/REPEAT1 - Collection Patterns

**What they are**:

- REPEAT: Zero or more occurrences of a pattern
- REPEAT1: One or more occurrences of a pattern

**Grammar Example**:

```json
// REPEAT - can be empty
{
  "directives": {
    "type": "REPEAT",
    "content": { "type": "SYMBOL", "name": "directive" }
  }
}

// REPEAT1 - must have at least one
{
  "definitions": {
    "type": "REPEAT1",
    "content": { "type": "SYMBOL", "name": "definition" }
  }
}
```

**What REPEAT/REPEAT1 generate in the Navigator API**:

- Collection navigator with array-like methods
- Individual item accessors by index
- Query and filtering operations
- Iterator protocol support

**Complete API for REPEAT pattern**:

For REPEAT and REPEAT1 patterns, we'll implement a rich API that combines:

- Full array method compatibility
- Iterator protocol support
- Query-focused operations
- Clear distinction between chain continuation and terminal methods

```
DirectivesNavigator «
  // Access individual items
  at(◆INDEX) → DirectiveNavigator?
  first() → DirectiveNavigator?
  last() → DirectiveNavigator?

  // Query operations
  whereName(◆NAME) → DirectiveNavigator?
  find(◆predicate) → DirectiveNavigator?
  filter(◆predicate) → DirectivesNavigator

  // Terminal operations
  values() → Directive[]
  map(◆FN) → ◆R[]
  count() → number
  isEmpty() → boolean

  // Iterator
  [Symbol.iterator]() → Iterator
»
```

**Usage Examples**:

```typescript
// REPEAT - can be empty
const deprecated = field.directives().whereName('deprecated');

// REPEAT1 - always has items
const firstDef = document.definitions().first(); // Never null

// Collection operations
const operationNames = document
  .definitions()
  .filter(d => d.is('operation_definition'))
  .map(d => d.ifOperationDefinition()?.name()?.text() ?? 'anonymous');
```

## FIELD - Named Fields Pattern

**What it is**: FIELD assigns names to child nodes, making them accessible via specific field names rather than positional access.

**Grammar Example**:

```json
{
  "operation_definition": {
    "type": "SEQ",
    "members": [
      {
        "type": "FIELD",
        "name": "operation_type",
        "content": { "type": "SYMBOL", "name": "operation_type" }
      },
      {
        "type": "FIELD",
        "name": "name",
        "content": {
          "type": "CHOICE",
          "members": [
            { "type": "SYMBOL", "name": "name" },
            { "type": "BLANK" }
          ]
        }
      }
    ]
  }
}
```

**What FIELD generates in the Navigator API**:

- Methods named after the field name
- Return type based on the field's content pattern
- Automatic camelCase conversion of field names

**Usage Example**:

```typescript
const operation = navigate(operationDef);
const type = operation.operationType(); // From FIELD name "operation_type"
const name = operation.name(); // Optional due to CHOICE with BLANK
```

## PREC/PREC_LEFT/PREC_RIGHT/PREC_DYNAMIC - Precedence Patterns

**What they are**: These patterns control parsing precedence and associativity but don't affect the AST structure.

**Grammar Example**:

```json
{
  "type": "PREC_RIGHT",
  "value": 0,
  "content": { "type": "SYMBOL", "name": "selection_set" }
}
```

**What precedence patterns generate**: Nothing! They're transparent to navigation - we generate API based on their content pattern.

## ALIAS - Node Renaming Pattern

**What it is**: ALIAS gives a node a different name in the AST than its grammar rule name.

**Grammar Example**:

```json
{
  "type": "ALIAS",
  "content": { "type": "SYMBOL", "name": "name" },
  "named": true,
  "value": "operation_name"
}
```

**What ALIAS generates**: The navigator uses the alias name instead of the original rule name.

## Terminal Patterns (STRING/PATTERN/TOKEN)

**What they are**: These create leaf nodes in the AST with no children.

**STRING**: Literal keywords like `"query"`, `"mutation"`, `"{"`, etc.

- Usually anonymous nodes (not accessible via navigation)
- Sometimes aliased to be named (e.g., operation types)

**PATTERN/TOKEN**: Regular expressions for identifiers, numbers, strings, etc.

- Always named nodes with `.text()` method
- No child navigation methods

**Complete API for terminal nodes**:

```
NameNavigator «
  value() → Name
  text() → string                      // The actual text content
  parent() → Navigator?

  // No child methods - it's a leaf node
»
```

**Usage Example**:

```typescript
const fieldName = field.name().text(); // "userId"
const intValue = argument.value().ifIntValue()?.text(); // "42"
```

# Examples

```typescript
// Chaining with continuation methods
const userFields = navigate(query)
  .selectionSet()
  .selections()
  .fields() // CollectionNavigator<Field>
  .whereName('user') // Navigator<Field> | null
  ?.selectionSet()
  .selections()
  .fields() // CollectionNavigator<Field>
  .filter(f => !f.alias().exists()) // CollectionNavigator<Field>
  .take(5); // CollectionNavigator<Field>

// Terminal methods end the chain
const fieldNames = userFields.map(f => f.name().text()); // string[]
const hasId = userFields.some(f => f.name().text() === 'id'); // boolean
const fields = userFields.values(); // Node.Field[]

// Using iterator
for (const field of userFields) {
  console.log(field.name().text());
}

// Complex query
const deprecatedFields = navigate(schema)
  .definitions()
  .whereType('object_type_definition')
  .flatMap(obj => obj.fieldsDefinition()?.fields() ?? [])
  .filter(f => f.directives().whereName('deprecated').exists())
  .map(f => ({
    typeName: f.ancestor('object_type_definition')?.name().text(),
    fieldName: f.name().text(),
    deprecationReason: f
      .directives()
      .whereName('deprecated')
      ?.arguments()
      ?.whereName('reason')
      ?.value()
      ?.text(),
  }));
```

At any point in the chain, `.value()` returns the current underlying node(s) with proper typing:

```typescript
// Single navigator
const field: Node.Field | null = navigate(selection)
  .ifField()
  ?.value();

// Collection navigator
const fields: Node.Field[] = navigate(selectionSet)
  .selections()
  .fields()
  .values();

// After filtering
const deprecatedFields: Node.Field[] = navigate(selectionSet)
  .selections()
  .fields()
  .where(f => f.directives().whereName('deprecated').exists())
  .values();

// Nullable chain
const aliasName: Node.Name | null = navigate(field)
  .alias() // Navigator<Node.Alias> | null
  ?.name() // Navigator<Node.Name> | null
  ?.value(); // Node.Name | null
```

Collection Operations

```typescript
// Find all deprecated fields
const deprecatedFields = schema
  .definitions()
  .filter(d => d.is('object_type_definition'))
  .flatMap(obj => obj.fieldsDefinition()?.fields().values() ?? [])
  .filter(f => f.directives()?.whereName('deprecated')?.exists() ?? false)
  .map(f => ({
    type: f.ancestor('object_type_definition')?.name().text(),
    field: f.name().text(),
  }));

// Get operation names
const operationNames = document
  .definitions()
  .filter(d => d.is('operation_definition'))
  .map(d => d.ifOperationDefinition()?.name()?.text() ?? 'anonymous');
```

Pattern Matching

```typescript
// Handle different selection types
const selectionInfo = selection.match({
  field: f => `Field: ${f.name().text()}`,
  fragment_spread: fs => `Fragment: ${fs.fragmentName().text()}`,
  inline_fragment: inf =>
    `Inline on ${inf.typeCondition()?.namedType().name().text()}`,
  _: () => 'Unknown selection',
});

// Type-safe value handling
const valueStr = value.match({
  string_value: s => s.text(),
  int_value: i => i.text(),
  boolean_value: b => b.text(),
  variable: v => `$${v.name().text()}`,
  _: () => 'complex value',
});
```
