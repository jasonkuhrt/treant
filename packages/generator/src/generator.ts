/**
 * Core generator implementation using Effect for all I/O
 */

import { createStreaming } from '@dprint/formatter';
import { FileSystem, Path } from '@effect/platform';
import { Grammar } from '@treant/grammar';
import { Console, Effect } from 'effect';
import { createHash } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { getNodeInterfaceName, getTypeGuardName } from './generators/type-helpers.js';
import * as TS from './lib/ts-syntax/$$.js';

// Import cursor generation system - now from grammar package
import { DEFAULT_CURSOR_CONFIG, generateCursorSystem } from './generators/cursor.js';

export interface GenerateOptions {
  grammarPath: string;
  nodeTypesPath: string;
  outputDir: string;
  nameOverride?: string;
}

/**
 * Write a file with optional formatting
 */
const writeFormattedFile = (
  path: string,
  content: string,
  formatter: any,
) =>
  Effect.gen(function*() {
    const fs = yield* FileSystem.FileSystem;

    try {
      const formatted = formatter.formatText(path, content);
      yield* fs.writeFileString(path, formatted || content);
    }
    catch (error) {
      yield* Console.warn(`Failed to format ${path}, writing unformatted`);
      yield* fs.writeFileString(path, content);
    }
  });

/**
 * Ensure a directory exists
 */
const ensureDir = (path: string) =>
  Effect.gen(function*() {
    const fs = yield* FileSystem.FileSystem;
    const pathApi = yield* Path.Path;

    // Check if exists
    const exists = yield* fs.exists(path).pipe(
      Effect.orElseSucceed(() => false),
    );

    if (!exists) {
      yield* fs.makeDirectory(path, { recursive: true });
    }
  });

/**
 * Load and parse JSON file
 */
const loadJson = <T = unknown>(path: string) =>
  Effect.gen(function*() {
    const fs = yield* FileSystem.FileSystem;
    const content = yield* fs.readFileString(path);
    return JSON.parse(content) as T;
  });

/**
 * Main generator function using Effect
 */
export const generate = (options: GenerateOptions) =>
  Effect.gen(function*() {
    const { grammarPath, nodeTypesPath, outputDir, nameOverride } = options;
    const fs = yield* FileSystem.FileSystem;
    const path = yield* Path.Path;

    // Load grammar and node types
    const grammarJson = yield* loadJson<Grammar.GrammarJson>(grammarPath);
    const nodeTypes = yield* loadJson<Grammar.NodeType[]>(nodeTypesPath);

    // Extract grammar name
    const grammarName = grammarJson.name;
    const grammarNamePascal = nameOverride || (grammarName === 'graphql' ? 'GraphQL' : TS.toPascalCase(grammarName));
    const grammarNamespaceExport = `TreeSitter${grammarNamePascal}`;

    // Initialize formatter

    // Try to load dprint config
    const dprintConfigPath = path.join(process.cwd(), 'dprint.json');
    const dprintConfig = yield* loadJson(dprintConfigPath).pipe(
      Effect.orElse(() =>
        Effect.succeed({
          typescript: {
            semiColons: 'prefer',
            quoteStyle: 'alwaysSingle',
            useBraces: 'whenNotSingleLine',
            bracePosition: 'sameLineUnlessHanging',
            singleBodyPosition: 'maintain',
            nextControlFlowPosition: 'nextLine',
            trailingCommas: 'onlyMultiLine',
            operatorPosition: 'nextLine',
            preferHanging: false,
            preferSingleLine: false,
          },
        })
      ),
    );

    const tsPluginUrl = (dprintConfig as any).plugins?.find((p: string) => p.includes('typescript'))
      || 'https://plugins.dprint.dev/typescript-0.93.0.wasm';

    // Cache directory for WASM files
    const cacheDir = `${homedir()}/.cache/treant`;
    const urlHash = createHash('md5').update(tsPluginUrl).digest('hex');
    const cacheFile = `${cacheDir}/formatter-${urlHash}.wasm`;

    // Ensure cache directory exists
    if (!existsSync(cacheDir)) {
      mkdirSync(cacheDir, { recursive: true });
    }

    // Load formatter with caching
    let formatter;
    if (existsSync(cacheFile)) {
      // Use cached WASM file
      const wasmBuffer = readFileSync(cacheFile);
      formatter = yield* Effect.promise(() => createStreaming(Promise.resolve(new Response(wasmBuffer))));
    }
    else {
      // Download and cache WASM file
      yield* Console.log('Downloading formatter (this is only needed once)...');
      const response = yield* Effect.promise(() => fetch(tsPluginUrl));
      const buffer = yield* Effect.promise(() => response.arrayBuffer());

      // Save to cache
      writeFileSync(cacheFile, Buffer.from(buffer));

      formatter = yield* Effect.promise(() => createStreaming(Promise.resolve(new Response(buffer))));
    }

    formatter.setConfig(
      {
        indentWidth: (dprintConfig as any).typescript?.indentWidth || 2,
        lineWidth: (dprintConfig as any).typescript?.lineWidth || 120,
      },
      (dprintConfig as any).typescript || {},
    );

    // Create output directories
    yield* ensureDir(path.join(outputDir, 'nodes'));
    yield* ensureDir(path.join(outputDir, 'cursor'));
    yield* ensureDir(path.join(outputDir, 'node', 'group'));

    // Extract node types
    const namedNodes = nodeTypes.filter(node => node.named === true);
    const anonymousNodes = nodeTypes.filter(node => node.named === false);

    // Generate node files
    yield* Effect.forEach(namedNodes, (node) =>
      Effect.gen(function*() {
        const nodeFile = generateNodeFile(node, grammarName, grammarNamePascal);
        const fileName = `${node.type}.ts`;
        const filePath = path.join(outputDir, 'nodes', fileName);
        yield* writeFormattedFile(filePath, nodeFile, formatter);
      }), { concurrency: 5 });

    // Generate other files

    // Generate types.ts
    const typesContent = generateTypesFile();
    yield* writeFormattedFile(path.join(outputDir, 'types.ts'), typesContent, formatter);

    // Generate utils.ts
    const utilsContent = generateUtilsFile(nodeTypes, grammarName);
    yield* writeFormattedFile(path.join(outputDir, 'utils.ts'), utilsContent, formatter);

    // Generate anonymous-nodes.ts
    const anonymousContent = generateAnonymousNodesFile(anonymousNodes, grammarName);
    yield* writeFormattedFile(path.join(outputDir, 'anonymous-nodes.ts'), anonymousContent, formatter);

    // Generate cursor system
    const grammarAnalysis = Grammar.Analysis.analyzeGrammar(nodeTypes, grammarJson);

    const cursorGeneratorConfig = {
      outputDir,
      formatter,
    };

    // Note: This would need to be refactored to use Effect as well
    yield* Effect.promise(() => generateCursorSystem(cursorGeneratorConfig, grammarAnalysis, DEFAULT_CURSOR_CONFIG));
  });

// Helper functions for generating different file types
function generateNodeFile(node: Grammar.NodeType, grammarName: string, grammarNamePascal: string): string {
  const nodeName = getNodeInterfaceName(node.type);
  const typeGuardName = getTypeGuardName(node.type);
  const lines: string[] = [];

  lines.push('import type { Node } from \'web-tree-sitter\';');

  // Import child types if needed
  if (node.children && node.children.types.length > 0) {
    const childImports = node.children.types
      .filter(t => t.named && t.type !== node.type)
      .map(t => `import type { ${getNodeInterfaceName(t.type)} } from './${t.type}.js';`)
      .filter((v, i, a) => a.indexOf(v) === i);

    if (childImports.length > 0) {
      childImports.forEach(imp => lines.push(imp));
    }
  }

  lines.push('');
  lines.push(`const TYPE = '${node.type}' as const;`);
  lines.push('');

  // Interface
  lines.push(`/**`);
  lines.push(` * Represents a ${node.type.replace(/_/g, ' ')} in the ${grammarName} AST.`);
  lines.push(` */`);
  lines.push(`export interface ${nodeName} extends Node {`);
  lines.push(`  type: typeof TYPE;`);
  lines.push('}');
  lines.push('');

  // Type guard
  lines.push(`/**`);
  lines.push(` * Type guard for ${nodeName}`);
  lines.push(` */`);
  lines.push(`export function ${typeGuardName}(node: unknown): node is ${nodeName} {`);
  lines.push(`  return (node as any)?.type === TYPE;`);
  lines.push('}');

  return lines.join('\n');
}

function generateTypesFile(): string {
  return `/**
 * Type definitions for tree-sitter AST nodes.
 * @generated
 */

export type { Node } from 'web-tree-sitter';
`;
}

function generateUtilsFile(nodeTypes: Grammar.NodeType[], grammarName: string): string {
  const namedNodes = nodeTypes.filter(n => n.named);

  return `/**
 * Utility functions for ${grammarName} AST traversal.
 * @generated
 */

import type * as WebTreeSitter from 'web-tree-sitter';

// Node type imports would go here...

export function findChildByType<T extends string>(
  node: WebTreeSitter.Node,
  type: T
): WebTreeSitter.Node | null {
  if (!node) return null;
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child?.type === type) {
      return child;
    }
  }
  return null;
}

export function findChildrenByType<T extends string>(
  node: WebTreeSitter.Node,
  type: T
): WebTreeSitter.Node[] {
  if (!node) return [];
  const results: WebTreeSitter.Node[] = [];
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child?.type === type) {
      results.push(child);
    }
  }
  return results;
}
`;
}

function generateAnonymousNodesFile(anonymousNodes: Grammar.NodeType[], grammarName: string): string {
  // Categorize anonymous nodes
  const punctuationNodes = anonymousNodes.filter(node => node.type.length === 1 && /^[^a-zA-Z0-9]$/.test(node.type));

  const operatorNodes = anonymousNodes.filter(node => node.type.length > 1 && /^[^a-zA-Z0-9]+$/.test(node.type));

  const keywordNodes = anonymousNodes.filter(node => /^[a-z]+$/.test(node.type));

  return `/**
 * Anonymous node types for the ${grammarName} AST.
 * @generated
 */

import type { Node } from 'web-tree-sitter';

// Punctuation types
export type PunctuationType = ${punctuationNodes.map(n => `'${n.type}'`).join(' | ') || 'never'};

// Operator types
export type OperatorType = ${operatorNodes.map(n => `'${n.type}'`).join(' | ') || 'never'};

// Keyword types
export type KeywordType = ${keywordNodes.map(n => `'${n.type}'`).join(' | ') || 'never'};
`;
}
