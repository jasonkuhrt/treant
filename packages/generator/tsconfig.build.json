{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "rootDir": "src",
    "noEmit": false,
    "outDir": "./build"
  },
  "include": ["src"]
}