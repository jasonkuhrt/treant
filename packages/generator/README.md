# @treant/generator

TypeScript SDK generator for Tree-sitter grammars.

Generates type-safe SDKs from Tree-sitter grammar definitions, providing:
- Type guards for every node type
- Type-safe navigation APIs
- Tree traversal utilities
- Full TypeScript type coverage

## Usage

```typescript
import { generate, emit, loadGrammar } from '@treant/generator';

// Load grammar
const grammar = await loadGrammar('./path/to/grammar/src');

// Generate SDK
const sdk = await generate({ grammar });

// Write to disk
await emit(sdk, './output');
```

## Features

- Generates complete TypeScript SDK from grammar.json
- Type guards for every node type
- Navigator API for type-safe tree traversal
- Cursor API for chainable navigation
- Parser wrapper with WASM integration
- Full type safety throughout

## License

MIT
