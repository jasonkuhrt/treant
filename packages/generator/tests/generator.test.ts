import type { Grammar } from '@treant/grammar';
import { expect, test } from 'vitest';
import { generate } from '../src/generator.js';

test('includes grammar artifacts in generated files when emitArtifacts is true', async () => {
  // Minimal grammar for testing
  const grammar: Grammar.BuiltGrammar = {
    grammarJson: {
      name: 'test_artifacts',
      rules: {
        source_file: {
          type: 'SYMBOL',
          name: 'expression',
        },
        expression: {
          type: 'STRING',
          value: 'hello',
        },
      },
    },

    nodeTypes: [
      {
        type: 'source_file',
        named: true,
      },
      {
        type: 'expression',
        named: true,
      },
    ],
    parserC: '', // Mock value for test
  };

  // Generate SDK with artifacts enabled
  const result = await generate({ grammar, emitArtifacts: true });

  expect(result.files.find(f => f.path === '__artifacts__/grammar.json')).toMatchSnapshot();
  expect(result.files.find(f => f.path === '__artifacts__/node-types.json')).toMatchSnapshot();
});

test('does not include artifacts by default', async () => {
  // Minimal grammar for testing
  const grammar: Grammar.BuiltGrammar = {
    grammarJson: {
      name: 'test_no_artifacts',
      rules: {
        source_file: {
          type: 'SYMBOL',
          name: 'expression',
        },
        expression: {
          type: 'STRING',
          value: 'hello',
        },
      },
    },

    nodeTypes: [
      {
        type: 'source_file',
        named: true,
      },
      {
        type: 'expression',
        named: true,
      },
    ],
    parserC: '', // Mock value for test
  };

  // Generate SDK without artifacts (default)
  const result = await generate({ grammar });

  expect(result.files.find(f => f.path === '__artifacts__/grammar.json')).toBeUndefined();
  expect(result.files.find(f => f.path === '__artifacts__/node-types.json')).toBeUndefined();
});
