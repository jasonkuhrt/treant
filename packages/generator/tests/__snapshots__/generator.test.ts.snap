// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`includes grammar artifacts in generated files when emitArtifacts is true 1`] = `
{
  "content": "{
  "name": "test_artifacts",
  "rules": {
    "source_file": {
      "type": "SYMBOL",
      "name": "expression"
    },
    "expression": {
      "type": "STRING",
      "value": "hello"
    }
  }
}",
  "path": "__artifacts__/grammar.json",
}
`;

exports[`includes grammar artifacts in generated files when emitArtifacts is true 2`] = `
{
  "content": "[
  {
    "type": "source_file",
    "named": true
  },
  {
    "type": "expression",
    "named": true
  }
]",
  "path": "__artifacts__/node-types.json",
}
`;
